File: src/leetcode/python/DiameterOfBinaryTree_543.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class DiameterOfBinaryTree_543(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        mx = [0, 0]

        def diameter(node):
            if not node:
                return 0

            l, r = diameter(node.left), diameter(node.right)
            mx[1] = max(l, r) + 1
            mx[0] = max(mx[0], l + r)
            return mx[1]

        diameter(root)
        return mx[0]

File: src/leetcode/java/DiameterOfBinaryTree_543.java
public class DiameterOfBinaryTree_543 {
  public int diameterOfBinaryTree(TreeNode root) {
    int[] max = new int[2];
    diameter(root, max);
    return max[0];
  }

  private int diameter(TreeNode node, int[] max) {
    if (node == null) return 0;
    int left = diameter(node.left, max);
    int right = diameter(node.right, max);
    max[0] = Math.max(max[0], left + right);
    max[1] = Math.max(left, right) + 1;
    return max[1];
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}
